cql3/restrictions/statement_restrictions.cc

```diff
@@ -1003,5 +1010,652 @@ ::shared_ptr<single_column_restriction> single_column_restriction::LIKE::apply_t
+                } else if (*opr.op == operator_type::GT || *opr.op == operator_type::GTE) {
+                    return false;
+                } else {
+                    throw std::logic_error("Unknown slice operator");
+                }
+            } else if (cmp > 0) {
+                if (*opr.op == operator_type::LT || *opr.op == operator_type::LTE) {
+                    return false;
+                } else if (*opr.op == operator_type::GT || *opr.op == operator_type::GTE) {
+                    return true;
+                } else {
+                    throw std::logic_error("Unknown slice operator");
+                }
+            }
+            // Otherwise, we don't know the LHS/RHS order, so check the next component.
+        }
+        // Getting here means LHS == RHS.
+        return *opr.op == operator_type::LTE || *opr.op == operator_type::GTE;
+    } else if (columns.size() == 1) {
+        auto lhs = get_value(columns[0], data);
+        if (!lhs) {
+            lhs = bytes(); // Compatible with old code, which feeds null to type comparators.
+        }
+        auto tup = dynamic_pointer_cast<tuples::value>(get_tuple(opr.rhs, data.options));
+        // Assume this is an external query WHERE (ck1)>(123), rather than an internal query WHERE col>(123),
+        // because internal queries have no reason to use single-element tuples.
+        //
+        // TODO: make the two cases distinguishable.
+        auto rhs = (tup && tup->size() == 1) ? tup->get_elements()[0]
+                : to_bytes_opt(opr.rhs->bind_and_get(data.options));
+        if (!rhs) {
+            return false;
+        }
+        return limits(*lhs, *opr.op, *rhs, *comparator(columns[0]));
+    } else {
+        throw std::logic_error("empty tuple on LHS of an inequality");
+    }
+}
+
+/// True iff collection (list, set, or map) contains value.
+bool contains(const data_value& collection, const raw_value_view& value) {
+    if (!value) {
+        return true; // Compatible with old code, which skips null terms in value comparisons.
+    }
+    auto col_type = static_pointer_cast<const collection_type_impl>(collection.type());
+    auto&& element_type = col_type->is_set() ? col_type->name_comparator() : col_type->value_comparator();
+    return with_linearized(*value, [&] (bytes_view val) {
+        auto exists_in = [&](auto&& range) {
+            auto found = std::find_if(range.begin(), range.end(), [&] (auto&& element) {
+                return element_type->compare(element.serialize_nonnull(), val) == 0;
+            });
+            return found != range.end();
+        };
+        if (col_type->is_list()) {
+            if (!exists_in(value_cast<list_type_impl::native_type>(collection))) {
+                return false;
+            }
+        } else if (col_type->is_set()) {
+            if (!exists_in(value_cast<set_type_impl::native_type>(collection))) {
+                return false;
+            }
+        } else {
```

Comment:
let's assert that `col_type->is_map()`. To provide a potential reason: introducing JSON document type to CQL has been mentioned several times already, so once it arrives, it will be easier to find all the places to fix.

Reply:
Do we still abort on assertion failures in release mode?  If so, then anything is better than an assert.

But I do agree that an extra check of some sort would be good here, just to validate the assumption it's a map.

Reply:
Done.